src/emulator/ui.ts

```diff
@@ -10,58 +14,99 @@ import { ALL_EXPERIMENTS, ExperimentName, isEnabled } from "../experiments";
 export interface EmulatorUIOptions {
   listen: ListenSpec[];
   projectId: string;
-  auto_download?: boolean;
 }
 
-export class EmulatorUI implements EmulatorInstance {
-  constructor(private args: EmulatorUIOptions) {}
+export class EmulatorUI extends ExpressBasedEmulator {
+  constructor(private args: EmulatorUIOptions) {
+    super({
+      listen: args.listen,
+    });
+  }
+
+  override async start(): Promise<void> {
+    await super.start();
+  }
 
-  start(): Promise<void> {
+  protected override async createExpressApp(): Promise<express.Express> {
     if (!EmulatorRegistry.isRunning(Emulators.HUB)) {
       throw new FirebaseError(
         `Cannot start ${Constants.description(Emulators.UI)} without ${Constants.description(
           Emulators.HUB,
         )}!`,
       );
     }
-    const { auto_download: autoDownload, projectId } = this.args;
-    const env: Partial<NodeJS.ProcessEnv> = {
-      LISTEN: JSON.stringify(ExpressBasedEmulator.listenOptionsFromSpecs(this.args.listen)),
-      GCLOUD_PROJECT: projectId,
-      [Constants.FIREBASE_EMULATOR_HUB]: EmulatorRegistry.url(Emulators.HUB).host,
-    };
+    const app = await super.createExpressApp();
+    const { projectId } = this.args;
+    const enabledExperiments: Array<ExperimentName> = (
+      Object.keys(ALL_EXPERIMENTS) as Array<ExperimentName>
+    ).filter((experimentName) => isEnabled(experimentName));
+    const emulatorGaSession = emulatorSession();
 
-    const session = emulatorSession();
-    if (session) {
-      env[Constants.FIREBASE_GA_SESSION] = JSON.stringify(session);
-    }
+    await downloadableEmulators.downloadIfNecessary(Emulators.UI);
+    const downloadDetails = downloadableEmulators.getDownloadDetails(Emulators.UI);
+    const webDir = path.join(downloadDetails.unzipDir!!, "client");
+
+    // Exposes the host and port of various emulators to facilitate accessing
+    // them using client SDKs. For features that involve multiple emulators or
+    // hard to accomplish using client SDKs, consider adding an API below.
+    app.get(
+      "/api/config",
+      this.jsonHandler(async () => {
+        const hubDiscoveryUrl = new URL(
+          `http://${EmulatorRegistry.url(Emulators.HUB).host}/emulators`,
+        );
+        const emulatorsRes = await fetch(hubDiscoveryUrl.toString());
+        const emulators = (await emulatorsRes.json()) as any;
```

Comment:
Nit: This is a bit code smelly and fragile - if emulatorsRes ever returns something other than a properly formatted JSON, this will crash with a "... on noniterable object" error

Might be worth refactoring this to fail cleaner, but not blocking

Reply:
If I'm not wrong, this can be done in-memory using `EmulatorRegistry` to directly get the hub instance. No need for HTTP at all.

Reply:
Nothing a leaky abstraction won't fix - PTAL

Reply:
Question: do we need EmulatorHub.PATH_EMULATORS anymore?

Reply:
> Question: do we need EmulatorHub.PATH_EMULATORS anymore?

Yes, that's part our public API and used for an SDK too